### ISSUE TITLE
Java types like `LinkedHashSet<E>` should allow nullable elements

### ISSUE THREAD

**Issue by sdeleuze** [uber/NullAway#1125]
In Spring Framework codebase with JSpecify mode enabled, when specifying
```java
Set<@Nullable String> result = CollectionUtils.newLinkedHashSet(patterns.length);
```
I get
```
 error: [NullAway] Cannot assign from type LinkedHashSet<String> to type Set<@Nullable String> due to mismatched nullability of type parameters
```

If I try
```
Set<@Nullable String> result = CollectionUtils.<@Nullable String>newLinkedHashSet(patterns.length);
```
I get
```
error: [NullAway] Type argument cannot be @Nullable, as method <E>newLinkedHashSet(int)'s type variable E is not @Nullable
```

`LinkedHashSet` Javadoc specifies "it permits null elements" so I think this is a bug.

**Comment by clivinn-shla81092** [comment]
Hello, we tried to solve the issue.

This is what we did:

Modify GenericsChecks to allow @Nullable type arguments for collection types that permit null elements, such as LinkedHashSet. This change addresses the issue where NullAway incorrectly reports errors for valid uses of @Nullable type arguments with collections that support null elements.

You can review changes in this commit: https://github.com/clivinn-shla81092/uber-NullAway-1125/commit/b0a8eac070ca6e77a3639a5ea34bc56c4b9443a5.


> [!CAUTION]
> Disclaimer: The concept of solution was created by AI and you should never copy paste this code before you check the correctness of generated code. Solution might not be complete, you should use this code as an inspiration only.

---

Latta AI seeks to solve problems in open source projects as part of its mission to support developers around the world. Learn more about our mission at https://latta.ai/ourmission . If you no longer want Latta AI to attempt solving issues on your repository, you can block this account.

**Comment by sdeleuze** [comment]
`CompletableFuture#completeAsync` also does not allow nullable generic type.

**Comment by msridhar** [comment]
I agree this is a bug.  Looking into it.

**Comment by msridhar** [comment]
@sdeleuze 

> In Spring Framework codebase with JSpecify mode enabled, when specifying
> 
> Set<@Nullable String> result = CollectionUtils.newLinkedHashSet(patterns.length);
> I get
> 
> ```
>  error: [NullAway] Cannot assign from type LinkedHashSet<String> to type Set<@Nullable String> due to mismatched nullability of type parameters
> ```

Hopefully some errors like this one will go away soon, when we land #1131 
> 
> If I try
> 
> ```
> Set<@Nullable String> result = CollectionUtils.<@Nullable String>newLinkedHashSet(patterns.length);
> ```
> 
> I get
> 
> ```
> error: [NullAway] Type argument cannot be @Nullable, as method <E>newLinkedHashSet(int)'s type variable E is not @Nullable
> ```
> 
> `LinkedHashSet` Javadoc specifies "it permits null elements" so I think this is a bug.

I think what's important here is the `CollectionUtils.newLinkedHashSet` method:

https://github.com/spring-projects/spring-framework/blob/3c4d5357232be289343d25a11f668be9a62a9f5a/spring-core/src/main/java/org/springframework/util/CollectionUtils.java#L138

In the version you are trying to check, the declaration would need to be changed to add a bound to the type variable:

```java
public static <E extends @Nullable Object> LinkedHashSet<E> newLinkedHashSet(int expectedSize) {
```
Do you still see the error after doing that?



> `CompletableFuture#completeAsync` also does not allow nullable generic type.

Can you give a small code example for this one?

**Comment by sdeleuze** [comment]
> In the version you are trying to check, the declaration would need to be changed to add a bound to the type variable:
> 
> public static <E extends @Nullable Object> LinkedHashSet<E> newLinkedHashSet(int expectedSize) {
>
> Do you still see the error after doing that?

Yes I do, I tried that and got the following error:
```
> Compilation failed; see the compiler output below.
  /Users/sdeleuze/workspace/spring-framework/spring-core/src/main/java/org/springframework/util/CollectionUtils.java:127: error: [NullAway] Cannot assign from type LinkedHashSet<String> to type Set<@Nullable String> due to mismatched nullability of type parameters
                Set<@Nullable String> result = CollectionUtils.newLinkedHashSet(8);
```

> > `CompletableFuture#completeAsync` also does not allow nullable generic type.
> 
> Can you give a small code example for this one?

```
public static void test(@Nullable String name) {
	CompletableFuture<@Nullable String> completableFuture = CompletableFuture.completedFuture(name);
}
```

Which generates the following error:
```
> Task :spring-core:compileJava
/Users/sdeleuze/workspace/spring-framework/spring-core/src/main/java/org/springframework/util/CollectionUtils.java:133: error: [NullAway] Cannot assign from type CompletableFuture<String> to type CompletableFuture<@Nullable String> due to mismatched nullability of type parameters
                CompletableFuture<@Nullable String> completableFuture = CompletableFuture.completedFuture(name);
                                                    ^
    (see http://t.uber.com/nullaway )
```

**Comment by msridhar** [comment]
For the `LinkedHashSet` example, this self-contained version compiles for me with NullAway 0.12.3:
```java
import org.jspecify.annotations.Nullable;
import org.jspecify.annotations.NullMarked;
import java.util.LinkedHashSet;
import java.util.Set;
@NullMarked
class Scratch {
  static <E extends @Nullable Object> LinkedHashSet<E> newLinkedHashSet() {
    return new LinkedHashSet<>();
  }
  public static void main(String[] args) {
    Set<@Nullable Object> s = Scratch.<@Nullable Object>newLinkedHashSet();
  }
}
```
@sdeleuze does it work for you if you both write `E extends @Nullable Object` and also pass the explicit `@Nullable Object` type argument at the call to `newLinkedHashSet`?

**Comment by msridhar** [comment]
The `CompletableFuture` example also works with an explicit type argument:
```java
  public static void test(@Nullable String name) {
    CompletableFuture<@Nullable String> completableFuture = CompletableFuture.<@Nullable String>completedFuture(name);
  }
```
So I think the issue here is in fact our lack of type argument inference for calls to generic methods (#1075) which will partially be addressed by #1131; I think #1131 should be enough to handle these examples.
